cpp/src/arrow/sparse_tensor.cc

```diff
@@ -60,17 +66,23 @@ class SparseTensorConverter<TYPE, SparseCOOIndex>
   using typename BaseClass::NumericTensorType;
   using typename BaseClass::value_type;
 
-  explicit SparseTensorConverter(const NumericTensorType& tensor) : BaseClass(tensor) {}
+  explicit SparseTensorConverter(const NumericTensorType& tensor,
+                                 const std::shared_ptr<DataType>& index_value_type)
+      : BaseClass(tensor, index_value_type) {}
 
+  template <typename IndexValueType>
   Status Convert() {
+    using c_index_value_type = typename IndexValueType::c_type;
+    const int64_t indices_elsize = sizeof(c_index_value_type);
+
     const int64_t ndim = tensor_.ndim();
     int64_t nonzero_count = -1;
     RETURN_NOT_OK(tensor_.CountNonZero(&nonzero_count));
 
     std::shared_ptr<Buffer> indices_buffer;
-    RETURN_NOT_OK(
-        AllocateBuffer(sizeof(int64_t) * ndim * nonzero_count, &indices_buffer));
-    int64_t* indices = reinterpret_cast<int64_t*>(indices_buffer->mutable_data());
+    RETURN_NOT_OK(AllocateBuffer(indices_elsize * ndim * nonzero_count, &indices_buffer));
```

Comment:
Not related to this PR, but at some point all buffer-allocating constructors should have a variant taking a MemoryPool argument.

Reply:
Filed https://issues.apache.org/jira/browse/ARROW-6503